### PR TITLE
Adding a "create user" to the api

### PR DESCRIPTION
**Pull request by nir-sopher**

### CODE REVIEW

File: docs/source/development/traffic_ops_api/v12/user.rst [hunk 1]
```diff
@@ -214,6 +214,153 @@ Users
...
+  +----------------------+--------+------------------------------------------------+
+  |``role``              | int    |                                                |
+  +----------------------+--------+------------------------------------------------+
+  |``roleName``          | string |                                                |
+  +----------------------+--------+------------------------------------------------+
+  |``stateOrProvince``   | string |                                                |
+  +----------------------+--------+------------------------------------------------+
+  |``uid``               | int    |                                                |
+  +----------------------+--------+------------------------------------------------+
+  | ``tenantId``         | int    | Owning tenant ID                               |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/user.rst]: Can you add tenant to this response?

File: traffic_ops/app/lib/API/User.pm [hunk 2]
```diff
@@ -226,6 +226,120 @@ sub update {
...
+	
+	if ( !defined($params->{role}) ) {
+		return $self->alert("role is required.");
+	}
+	
+	#setting tenant_id to the user's tenant if tenant is not set. TODO(nirs): remove when tenancy is no longer optional in the API
+	my $tenant_id = exists($params->{tenantId}) ? $params->{tenantId} :  $self->current_user_tenant();
+
+	my $values = {
+		address_line1 			=> defined_or_default($params->{addressLine1}, ""),
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: I don't think you need to do all these defined_or_default()

it's fine if they leave out addressLine1 which will just set address_line1 to undef.

File: traffic_ops/app/lib/API/User.pm [hunk 3]
```diff
@@ -226,6 +226,120 @@ sub update {
...
+		address_line2 			=> defined_or_default($params->{addressLine2}, ""),
+		city 				=> defined_or_default($params->{city}, ""),
+		company 			=> defined_or_default($params->{company}, ""),
+		country 			=> defined_or_default($params->{country}, ""),
+		email 				=> $params->{email},
+		full_name 			=> $params->{fullName},
+		new_user            		=> defined_or_default($params->{newUser}, 0),
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: just set this to

new_user            		=> \0, # false

it's confusing, i know but "new_user" is used for users that were created via a registrationSent.

File: docs/source/development/traffic_ops_api/v12/user.rst (hunk 1 above)
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/user.rst]: nevermind, i guess this is checked in the is_valid() function.

File: traffic_ops/app/lib/API/User.pm [hunk 4]
```diff
@@ -226,6 +226,120 @@ sub update {
...
+		$response->{phoneNumber} 		= $rs->phone_number;
+		$response->{postalCode} 		= $rs->postal_code;
+		$response->{publicSshKey} 		= $rs->public_ssh_key;
+		$response->{registrationSent} 		= \$rs->registration_sent;
+		$response->{role} 			= $rs->role->id;
+		$response->{roleName} 			= $rs->role->name;
+		$response->{stateOrProvince} 		= $rs->state_or_province;
+		$response->{uid} 			= $rs->uid;
+		$response->{username} 			= $rs->username;
+		$response->{tenantId} 			= $rs->tenant_id;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: can you return the tenant name as well?

File: traffic_ops/app/lib/API/User.pm [hunk 5]
```diff
@@ -218,6 +218,115 @@ sub update {
...
+	if ($existing) {
+		return $self->alert("A user with username \"$name\" already exists.");
+	}
+
+
+	if ( !defined($params->{fullName}) ) {
+		return $self->alert("full-name is required.");
+	}
+
+	if ( !defined($params->{email}) ) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: nevermind, i guess email uniqueness is checked in the is_valid() function

**Comment by nir-sopher** [comment]
Indeed

File: docs/source/development/traffic_ops_api/v12/user.rst (hunk 1 above)
Comment (nir-sopher) [inline docs/source/development/traffic_ops_api/v12/user.rst]: Ori plans to do so in a separate PR or on the merge with pr 434

File: traffic_ops/app/lib/API/User.pm (hunk 5 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: Indeed

File: traffic_ops/app/lib/API/User.pm (hunk 4 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: Ori plans to add it in a separate CL

File: traffic_ops/app/lib/API/User.pm (hunk 2 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: done

File: traffic_ops/app/lib/API/User.pm (hunk 3 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: Done. "0" an not "\0" as setting it to "\0" failed.
Will need to look further into it to understand why

**Comment by mitchell852** [comment]
you might want to run your code thru the perl formatter. I see some formatting errors. Here's more info about that:

https://github.com/apache/incubator-trafficcontrol/blob/master/CONTRIBUTING.md#code-formatting

http://trafficcontrol.apache.org/docs/latest/development/traffic_ops.html#perl-formatting-conventions

File: traffic_ops/app/lib/API/User.pm [hunk 6]
```diff
@@ -226,6 +226,120 @@ sub update {
...
+	my $self = shift;
+	my $params = $self->req->json;
+	
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	my $name = $params->{username};
+	if ( !defined($name) ) {
+		return $self->alert("Username is required.");
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: you check for username, fullName and email in the is_valid method. so you don't have to do that here.

File: traffic_ops/app/t/api/1.2/user_admin.t [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	Test::TestHelper->teardown( $schema, 'Role' );
+	Test::TestHelper->teardown( $schema, 'TmUser' );
+
+	Test::TestHelper->load_core_data($schema);
+
+	my $tenant_id = $schema->resultset('TmUser')->find( { username => $login_user } )->get_column('tenant_id');
+	my $tenant_name = defined ($tenant_id) ? $schema->resultset('Tenant')->find( { id => $tenant_id } )->get_column('name') : "null";
+
+	# Verify the user
+	ok my $user = $schema->resultset('TmUser')->find( { username => $login_user } ), 'Does the portal user exist?';
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/user_admin.t]: Can you update this message? not sure what portal user has to do with this.

File: traffic_ops/app/t/api/1.2/user_admin.t [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	my $addedUserName = "user1";
+	my $addedUserEmail = "abc\@z.com";
+
+	ok $t->post_ok('/api/1.2/users' => {Accept => 'application/json'} => json => {
+        	"username" => $addedUserName, "fullName"=>"full name", "email" => $addedUserEmail, "localPassword" => "pass", "confirmLocalPassword"=> "pass", "role" => 4 })
+        	->status_is(200)->or( sub { diag $t->tx->res->content->asset->{content}; } )
+		->json_is( "/response/username" =>  $addedUserName )
+		->json_is( "/response/email" =>  $addedUserEmail)
+		->json_is( "/response/tenantId" =>  $tenant_id) #tenant Id not set - getting the tenant id from the user
+        	    , 'Failed adding user?';
```
Comment (mitchell852) [inline traffic_ops/app/t/api/1.2/user_admin.t]: this test message doesn't seem right either. Maybe "Successfully added user?" would be better.

File: traffic_ops/app/lib/API/User.pm (hunk 6 above)
Comment (nir-sopher) [inline traffic_ops/app/lib/API/User.pm]: I'll remove the checks for full-name and email.
As "username" value is use for searching the DB later on, I prefer to validate it.

File: traffic_ops/app/t/api/1.2/user_admin.t (hunk 7 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/user_admin.t]: Done

File: traffic_ops/app/t/api/1.2/user_admin.t (hunk 8 above)
Comment (nir-sopher) [inline traffic_ops/app/t/api/1.2/user_admin.t]: Done

**Comment by mitchell852** [comment]
@nir-sopher - can you fix this PR? "This branch cannot be rebased safely" -  i'd like to get it merged if possible

File: traffic_ops/app/lib/API/User.pm (hunk 6 above)
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: I don't understand. fullname, email and username are validated  here:

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/API/User.pm#L463

also,  you should probably add role to line 463

File: traffic_ops/app/lib/API/User.pm [hunk 9]
```diff
@@ -229,6 +229,112 @@ sub update {
...
+
+	if ( !defined($params->{localPassword}) ) {
+		return $self->alert("local-password is required.");
+	}
+
+	if ( !defined($params->{confirmLocalPassword}) ) {
+		return $self->alert("confirm-local-password is required.");
+	}
+
+	if ($params->{localPassword} ne $params->{confirmLocalPassword}){
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: this is not necessary, it is validated in the is_valid method

File: traffic_ops/app/lib/API/User.pm [hunk 10]
```diff
@@ -229,6 +229,112 @@ sub update {
...
+		return $self->alert("local-password and confirmed-local-password mismatch.");
+	}
+	
+	if ( !defined($params->{role}) ) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/User.pm]: if you add 'role' to line 463, you can take this out as well....

**Comment by nir-sopher** [comment]
Rebase was done a PR review changes were made, except for a test for "username" in the beginning of the create function. The test is done as the value is in use by the function logic.

**Comment by nir-sopher** [comment]
The change of the "role" broke the UT and was reverted.
It changes current behavior of user-update.
I prefer not to make such changes in this PR

**Comment by mitchell852** [comment]
yes, there is a UT failure:

t/user.t                                      (Wstat: 256 Tests: 21 Failed: 1)
  Failed test:  5
  Non-zero exit status: 1

but i'm pretty sure this was fixed in master so i'll pull this in and if it's still broken, i can fix it.

it  only changes the behavior of user-update in the sense that role is now checked which in my opinion is a very valid check. you should not be able to update a user and leave out a required field - role.
